assets/js/omniSelectEngine.js

```diff
@@ -0,0 +1,394 @@
+    $("body .select2-container").remove();
+  }
+
+  // Set up select content based on selected element.
+  function reinitOmniSearchForType(type, data) {
+    switch (type) {
+      case "task": // ToDo.
+        reinitOmniSearchWithData([
+          {
+            id: "",
+            text: "action",
+            children: [
+              {
+                id: data.id,
+                text: "Go To",
+                type: "taskAction",
+                action: "goto",
+              },
+              {
+                id: data.id,
+                text: "Log Time",
+                type: "taskAction",
+                action: "logtime",
+              },
+            ],
+          },
+        ]);
+        break;
+      case "project": // Project.
+        reinitOmniSearchWithData([
+          {
+            id: "",
+            text: "action",
+            children: [
+              {
+                id: data.id,
+                text: "Go To",
+                type: "projectAction",
+                action: "goto",
+              },
+              {
+                id: data.id,
+                text: "Create new To-Do",
+                type: "projectAction",
+                action: "createnew",
+              },
+            ],
+          },
+        ]);
+        break;
+    }
+  }
+  // Set data and refresh select2.
+  function reinitOmniSearchWithData(data) {
+    omniSelectElement
+      .select2("destroy")
+      .empty()
+      .select2({
+        data: data,
+        templateResult: function (data, container) {
+          // Setup custom options with icon and data values.
+          var $state = $(
+            '<div class="select2-results__option-container">' +
+              (data.type
+                ? "<span class='" + elementIcons[data.type] + "'></span>"
+                : "") +
+              data.text +
+              "</div>",
+          );
+          if (data.tags) {
+            $(container).attr("data-tags", data.tags);
+          }
+          if (data.type) {
+            $(container).attr("data-type", data.type);
+          }
+          if (data.client) {
+            $(container).attr("data-client", data.client);
+          }
+          return $state;
+        },
+        matcher: matcher,
+      })
+      .trigger("change")
+      .select2("open");
+
+    // Listens for escape key to close omni-search overlay when select2 is in focus.
+    $("body .select2-search__field").on("keydown", (e) => {
+      var searchFieldInputLength = $("body .select2-search__field").val()
+        .length;
+      var hasSelection =
+        $(".select2.select2-container").attr("data-visible-selected") &&
+        $(".select2.select2-container").attr("data-visible-selected").length >
+          0 &&
+        searchFieldInputLength == 0;
+      switch (e.keyCode) {
+        case 8:
+          if (hasSelection) {
+            destroyOmniSearch();
+            initOmniSearch();
+          }
+
+          break;
+        case 27:
```

Comment:
Consider declaring `const`s for these key codes to make the code easier to understand (in lines 30–33 you add comments to tell what “27” means, but contants will make the code much easier to understand).

Reply:
I did this!